### PR TITLE
[BugFix] Fix 'NPE: The URI scheme of endpointOverride must not be null'

### PR DESCRIPTION
**Pull request by lipenglin**
When there is no 'Scheme' in the 'aws.s3.endpoint' catalog properties, we
should consider this as an invalid configuration, and do not set the
'aws SdkClient endpointOverride' property.

refs:
software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder#endpointOverride
```
  public final B endpointOverride(URI endpointOverride) {
      if (endpointOverride == null) {
          this.clientConfiguration.option(SdkClientOption.ENDPOINT, (Object)null);
          this.clientConfiguration.option(SdkClientOption.ENDPOINT_OVERRIDDEN, false);
      } else {
          Validate.paramNotNull(endpointOverride.getScheme(), "The URI scheme of endpointOverride");
          this.clientConfiguration.option(SdkClientOption.ENDPOINT, endpointOverride);
          this.clientConfiguration.option(SdkClientOption.ENDPOINT_OVERRIDDEN, true);
      }

      return this.thisBuilder();
  }
```

exception_stack:
```
java.lang.NullPointerException: The URI scheme of endpointOverride must not be null.
        at software.amazon.awssdk.utils.Validate.paramNotNull(Validate.java:156) ~[bundle-2.17.257.jar:?]
        at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.endpointOverride(SdkDefaultClientBuilder.java:439) ~[bundle-2.17.257.jar:?]
        at com.starrocks.connector.iceberg.IcebergAwsClientFactory.s3(IcebergAwsClientFactory.java:155) ~[starrocks-fe.jar:?]
        at org.apache.iceberg.aws.s3.S3FileIO.client(S3FileIO.java:326) ~[iceberg-aws-1.2.1.jar:?]
        at org.apache.iceberg.aws.s3.S3FileIO.newInputFile(S3FileIO.java:124) ~[iceberg-aws-1.2.1.jar:?]
        at org.apache.iceberg.io.ResolvingFileIO.newInputFile(ResolvingFileIO.java:61) ~[iceberg-core-1.2.1.jar:?]
        at com.starrocks.connector.iceberg.io.IcebergCachingFileIO.newInputFile(IcebergCachingFileIO.java:133) ~[starrocks-fe.jar:?]
```

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by Smith-Cruise — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java [hunk 1]
```diff
@@ -151,8 +151,9 @@ public S3Client s3() {
             s3ClientBuilder.region(Region.of(s3Region));
         }
 
-        if (!s3Endpoint.isEmpty()) {
-            s3ClientBuilder.endpointOverride(URI.create(s3Endpoint));
+        URI uri;
+        if (!s3Endpoint.isEmpty() && (uri = URI.create(s3Endpoint)).getScheme() != null) {
```
Comment (Smith-Cruise) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:155]: add comments better

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java [hunk 2]
```diff
@@ -175,8 +176,9 @@ public GlueClient glue() {
             glueClientBuilder.region(Region.of(glueRegion));
         }
 
-        if (!glueEndpoint.isEmpty()) {
-            glueClientBuilder.endpointOverride(URI.create(glueEndpoint));
+        URI uri;
+        if (!glueEndpoint.isEmpty() && (uri = URI.create(glueEndpoint)).getScheme() != null) {
```
Comment (Smith-Cruise) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:180]: add comments better

Comment (lipenglin) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:184]: done

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java (hunk 1 above)
Comment (lipenglin) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:157]: done

**Review by lipenglin — COMMENTED** [review]
Thanks for comments!

**Review by imay — DISMISSED** [review]

**Review by stephen-shelby — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java (hunk 1 above)
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:157]: if it is a valid URI, just lack of scheme, can we assume it is `HTTPS`?

e.g. if the override endpoint provided is `abcbucket.s3.us-east.amazonaws.com`, we will safely assume it is `https://abcbucket.s3.us-east.amazonaws.com`?

make the code robust while keep easy to use.

Comment (lipenglin) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java:157]: It is not necessary, `endpointOverride` can be any valid S3 endpoint, including endpoints that use different protocols. Typically, AWS S3 services use the HTTPS protocol for secure data transmission, so in most cases, endpointOverride should also use the HTTPS protocol.

But if we can assume it is HTTPS, then in the `software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.endpointOverride` method, it should be done in this way that does not just check if the scheme is missing and just throw a NPE exception.

**Review by Smith-Cruise — APPROVED** [review]

**Comment by lipenglin** [comment]
If the endpoint is not empty but does not have a scheme, we choose to skip configuring the endpointOverride, but we should log a warning message to inform the user.
This approach can effectively fix the scenario where a large number of configurations have been made `without a scheme for the endpoint`. It allows users to have sufficient time to gradually fix their configurations, not affecting use, while also informing them of their wrong configuration through a warning message.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java (hunk 1 above)
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java]: This is just a java implementation of the sdk behavior. We can tune it to make it easy to use if necessary.

Ignoring the endpoint is better than NPE. But does it good though?

With this patch,  the endpoint override is ignored it in silence, user will not know the config item is ignored until something is not working. throwing ddlexception would be better than just ignoring.

As for using https if the scheme empty, it is the default behavior of cpp version of aws sdk, check the following code. I don't think it is not something we should not do, it is just how we define the behavior in such case.

https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/client/ClientConfiguration.h#L107
```
             * Http scheme to use. E.g. Http or Https. Default HTTPS
             */
            Aws::Http::Scheme scheme;

```

https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/endpoint/BuiltInParameters.cpp#L63
```
        if (!config.endpointOverride.empty()) {
            OverrideEndpoint(config.endpointOverride, config.scheme);
```

https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/endpoint/BuiltInParameters.cpp#L15
```
    void BuiltInParameters::OverrideEndpoint(const Aws::String& endpoint, const Aws::Http::Scheme& scheme)
    {
        static const Aws::String SDK_ENDPOINT = "Endpoint";

        if (endpoint.compare(0, 7, "http://") == 0 || endpoint.compare(0, 8, "https://") == 0)
        {
            SetStringParameter(SDK_ENDPOINT, endpoint);
        }
        else
        {
            SetStringParameter(SDK_ENDPOINT, Aws::String(Aws::Http::SchemeMapper::ToString(scheme)) + "://" + endpoint);
        }
    }

```

Comment (lipenglin) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergAwsClientFactory.java]: done.
Got it! This is more robust.
Thanks for comments!

**Review by kevincai — APPROVED** [review]
thanks! it would be prefect if we can add ut to cover this behavior.  `S3Utils.getUrl` might be leveraged to test the endpoint override behavior.

**Comment by lipenglin** [comment]
> thanks! it would be prefect if we can add ut to cover this behavior. `S3Utils.getUrl` might be leveraged to test the endpoint override behavior.

done.

**Review by kevincai — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by imay** [comment]
@mergify rebase main

**Pull request merged**